### PR TITLE
[SYCL] Disambiguate between load functions for multi_ptr

### PR DESCRIPTION
**Pull request by lbushi25**
This PR disambiguates between the two load functions for multi_ptr type in the case that DataT is const-qualified. For example, refer here https://github.com/intel/llvm/issues/12501 for a compilation error caused by this ambiguity. Patch suggested by @steffenlarsen.

### CODE REVIEW

**Comment by aelovikov-intel** [comment]
> For example, refer here https://github.com/intel/llvm/issues/12501 for an example of a compilation error

That needs to be fixed :)

**Comment by aelovikov-intel** [comment]
This needs a test added as part of this PR.

**Comment by lbushi25** [comment]
> This needs a test added as part of this PR.

Done, please have a look!

File: sycl/test/regression/vector_const_data.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// RUN: %clangxx -fsycl %s -fsyntax-only
```
Comment (aelovikov-intel) [inline sycl/test/regression/vector_const_data.h:1]: Why is this `.h` instead of `.cpp`?

**Comment by aelovikov-intel** [comment]
> > For example, refer here #12501 for an example of a compilation error
> 
> That needs to be fixed :)

That (double "for example") is still unaddressed.

File: sycl/test/regression/vector_const_data.h (hunk 1 above)
Comment (lbushi25) [reply sycl/test/regression/vector_const_data.h:1]: Just noticed that, changed it!

**Comment by lbushi25** [comment]
> > > For example, refer here #12501 for an example of a compilation error
> > 
> > 
> > That needs to be fixed :)
> 
> That (double "for example") is still unaddressed.

Good catch, though you were talking about the issue!
Addressed it.

File: sycl/test/regression/vector_const_data.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// sycl::vec should even be allowed to be instantiated with a const-qualified
+// type. If that discussion eventually resolves to the answer no, then this test
+// will probably fail compilation at line 9 and will need to be deleted.
+// The reason it is a test for now is that it verifies that overload resolution
+// succeeds for the function template member load of sycl::vec when instantiated
+// with a const-qualified type.
+int main() {
+  sycl::queue Q;
+  sycl::vec<int, 16> my_vec{};
+  sycl::vec<const int, 16> my_const_vec{};
```
Comment (aelovikov-intel) [inline sycl/test/regression/vector_const_data.cpp:13]: Are you sure that fails without your fix?

File: sycl/test/regression/vector_const_data.cpp [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+#include <sycl/sycl.hpp>
+// TODO: There are some spec discussions currently on hold about whether
+// sycl::vec should even be allowed to be instantiated with a const-qualified
+// type. If that discussion eventually resolves to the answer no, then this test
+// will probably fail compilation at line 13 and will need to be deleted.
+// The reason it is a test for now is that it verifies that overload resolution
+// succeeds for the function template member load of sycl::vec when instantiated
+// with a const-qualified type.
+int main() {
+  sycl::queue Q;
```
Comment (aelovikov-intel) [inline sycl/test/regression/vector_const_data.cpp:11]:
```suggestion
```

**Review by aelovikov-intel — DISMISSED** [review]

**Comment by steffenlarsen** [comment]
See https://github.com/KhronosGroup/SYCL-Docs/issues/540.

**Comment by gmlueck** [comment]
In general, we should not implement DPC++ differently from the spec.  This is bad for the ecosystem because it leads to non-portable application code that works only on DPC++.  If we think there is a bug in the spec, we should work towards getting the spec fixed first.

In this case, does the problem only occur when a `vec` is declared with a `const` data type?  If an application wants to create a "const vector", it seems like a better way to do this is to make the entire `vec` const like:

```
const sycl::vec<int, 16> my_vec{};
```

If there is an immediate user request around this feature, can we ask them to use this construct instead?

I'm not opposed to changing the specification for `vec::load`, but I think we should get a PR to change the spec first.  @AerialMantis, you were looking at all the `multi_ptr` APIs a little while ago.  Is there a precedent we should follow here?

**Comment by gmlueck** [comment]
> > In this case, does the problem only occur when a vec is declared with a const data type?
> 
> Partially! That is at least the reason why we found the problem. The current issue is that in DPC++ we tried to allow the conversion "implicitly" by defining it for both const and non-const `DataT`, but the way it is currently implemented becomes a problem if `DataT` is `const`, as it results in two declarations taking the same arguments.
> 
> If we follow SYCL 2020 to the letter, `load` doesn't work with `multi_ptr<DataT, ...>` if `DataT` isn't `const`, as illustrated in https://godbolt.org/z/dTdc6K7a9. That is, unless the user explicitly converts the multi-pointer to a `multi_ptr<const DataT, ...>` first.

We really should be following the spec to the letter, or requesting changes to the spec if we find problems.  We should not just define our own APIs because this will lead to application code that is not portable to other implementations of SYCL.

I added a comment to KhronosGroup/SYCL-Docs#540 with a different option for fixing this.  Let's wait for that to resolve before changing the code.

I also noticed that a few lines down we added overloads of `vec::load` and `vec::store` that take `accessor` as a parameter.  These are also not in the spec, but I opened KhronosGroup/SYCL-Docs#547 suggesting that they be added.  However, the order of the proposed template parameters is slightly different from what we have.  We should keep an eye on this issue and adjust our implementation if necessary.

Finally, I noticed that a few lines further down, we have `vec::ConvertToDataT` as a public member function.  This seems like a mistake in our code.  I think this is an implementation detail, so it should be private.

**Pull request closed without merging**